net/src/memory_server.rs

```diff
@@ -253,140 +253,248 @@ impl InMemoryServer {
+    }
+
     /// process a message sent by a node to the "network"
     pub fn serve(&mut self, data: Protocol) -> NetResult<()> {
         self.log
             .d(&format!(">>>> '{}' recv: {:?}", self.name.clone(), data));
-        if let Ok(json_msg) = JsonProtocol::try_from(&data) {
-            // Note: use same order as the enum
-            match json_msg {
-                JsonProtocol::SuccessResult(msg) => {
-                    // Relay directly the SuccessResult message
-                    self.priv_send_one(
-                        &msg.dna_address,
-                        &msg.to_agent_id,
-                        JsonProtocol::SuccessResult(msg.clone()).into(),
-                    )?;
-                }
-                JsonProtocol::FailureResult(msg) => {
-                    // Check if its a response to our own request
-                    let maybe_bucket_id = self.priv_check_request(&msg.request_id);
-                    if let Some(_) = maybe_bucket_id {
-                        self.log.d(&format!(
-                            "---- '{}' internal request failed: {:?}",
-                            self.name.clone(),
-                            msg.clone(),
-                        ));
-                        return Ok(());
-                    }
-                    // If not, relay the FailureResult message to receipient
-                    self.priv_send_one(
-                        &msg.dna_address,
-                        &msg.to_agent_id,
-                        JsonProtocol::FailureResult(msg.clone()).into(),
-                    )?;
-                }
-                JsonProtocol::TrackDna(msg) => {
-                    // Check if we are already tracking this dna for this agent
-                    let bucket_id = into_bucket_id(&msg.dna_address, &msg.agent_id);
-                    if self.trackdna_book.contains(&bucket_id) {
-                        self.log.e(&format!(
-                            "({}) ##### DNA already tracked: {}",
-                            self.name.clone(),
-                            bucket_id
-                        ));
-                        return Ok(());
-                    }
-                    self.trackdna_book.insert(bucket_id);
-                    // Notify all Peers connected to this DNA of a new Peer connection.
-                    self.priv_send_all(
-                        &msg.dna_address.clone(),
-                        JsonProtocol::PeerConnected(PeerData {
-                            agent_id: msg.agent_id.clone(),
-                        })
-                        .into(),
-                    )?;
-                    // Request all data lists from this agent
-                    self.priv_request_all_lists(&msg.dna_address, &msg.agent_id);
-                }
-
-                JsonProtocol::SendMessage(msg) => {
-                    self.priv_serve_SendMessage(&msg)?;
+        // serve only JsonProtocol
```

Comment:
Only change here is dealing with the Error case first so we dont have to indent the rest of the code.